Mathlib/Order/KonigLemma.lean

```diff
@@ -0,0 +1,147 @@
+is covered by only finitely many others, and `b` is an element with infinitely many things above it,
+then there is a sequence starting with `b` in which each element is covered by the next. -/
+theorem exists_seq_covby_of_forall_covby_finite (hfin : ∀ (a : α), {x | a ⋖ x}.Finite)
+    (hb : (Ici b).Infinite) : ∃ f : ℕ → α, f 0 = b ∧ ∀ i, f i ⋖ f (i+1) :=
+  let h := fun a : {a : α // (Ici a).Infinite} ↦
+    exists_covby_infinite_Ici_of_infinite_Ici a.2 (hfin a)
+  let ks : ℕ → {a : α // (Ici a).Infinite} := Nat.rec ⟨b, hb⟩ fun _ a ↦ ⟨_, (h a).choose_spec.2⟩
+  ⟨fun i ↦ (ks i).1, by simp [ks], fun i ↦ by simpa using (h (ks i)).choose_spec.1⟩
+
+/-- The sequence given by Kőnig's lemma as an order embedding -/
+theorem exists_orderEmbedding_covby_of_forall_covby_finite (hfin : ∀ (a : α), {x | a ⋖ x}.Finite)
+    (hb : (Ici b).Infinite) : ∃ f : ℕ ↪o α, f 0 = b ∧ ∀ i, f i ⋖ f (i+1) := by
+  obtain ⟨f, hf⟩ := exists_seq_covby_of_forall_covby_finite hfin hb
+  exact ⟨OrderEmbedding.ofStrictMono f (strictMono_nat_of_lt_succ (fun i ↦ (hf.2 i).lt)), hf⟩
+
+/-- A version of Kőnig's lemma where the sequence starts at the minimum of an infinite order.  -/
+theorem exists_orderEmbedding_covby_of_forall_covby_finite_of_bot [OrderBot α] [Infinite α]
+    (hfin : ∀ (a : α), {x | a ⋖ x}.Finite) : ∃ f : ℕ ↪o α, f 0 = ⊥ ∧ ∀ i, f i ⋖ f (i+1) :=
+  exists_orderEmbedding_covby_of_forall_covby_finite hfin (by simpa using infinite_univ)
+
+theorem GradeMinOrder.exists_nat_orderEmbedding_of_forall_covby_finite
+    [GradeMinOrder ℕ α] [OrderBot α] [Infinite α] (hfin : ∀ (a : α), {x | a ⋖ x}.Finite) :
+    ∃ f : ℕ ↪o α, f 0 = ⊥ ∧ (∀ i, f i ⋖ f (i+1)) ∧ ∀ i, grade ℕ (f i) = i := by
+  obtain ⟨f, h0, hf⟩ := exists_orderEmbedding_covby_of_forall_covby_finite_of_bot hfin
+  refine ⟨f, h0, hf, fun i ↦ ?_⟩
+  induction' i with i ih; simp [h0]
+  simpa [Nat.covBy_iff_succ_eq, ih, eq_comm] using CovBy.grade ℕ <| hf i
+
+end Sequence
+
+section Graded
+
+/-- A formulation of Kőnig's infinity lemma, useful in applications.
+Given a sequence `α 0, α 1, ...` of nonempty types with `α 0` a subsingleton,
+and a well-behaved family of projections `π : α j → α i` for all `i ≤ j`,
+if each term in each `α i` is the projection of only finitely many terms in `α (i+1)`,
+then we can find a sequence `(f 0 : α 0), (f 1 : α 1), ...`
+where `f i` is the projection of `f j` for all `i ≤ j`.
+
+In a typical application, the `α i` are function types with increasingly large domains,
+and `π hij (f : α j)` is the restriction of the domain of `f` to that of `α i`.
+In this case, the sequence given by the lemma is essentially a function whose domain
+is the limit of the `α i`. -/
+theorem exists_seq_forall_proj_of_forall_finite {α : ℕ → Type*} [Subsingleton (α 0)]
+    [∀ i, Nonempty (α i)] (π : {i j : ℕ} → (hij : i ≤ j) → α j → α i)
+    (π_trans : ∀ ⦃i j k⦄ (hij : i ≤ j) (hjk : j ≤ k) a, (π hij) (π hjk a) = π (hij.trans hjk) a)
+    (π_refl : ∀ ⦃i⦄ (a : α i), π rfl.le a = a)
```

Comment:
I think it would be a bit nicer to follow the category theory library here and state preservation of identity before preservation of composition
```suggestion
    (π_refl : ∀ ⦃i⦄ (a : α i), π rfl.le a = a)
    (π_trans : ∀ ⦃i j k⦄ (hij : i ≤ j) (hjk : j ≤ k) a, (π hij) (π hjk a) = π (hij.trans hjk) a)
```